### ISSUE TITLE
mutation_variables map(string)

### ISSUE THREAD

**Issue by robwolff3** [sullivtr/terraform-provider-graphql#34]
Hey, Love the provider you've made here. Super useful for the graph api I'm trying to manage resources for. 

One issue I think I'm running into. When trying to pass various variable types through to my graphql mutation I seem to be getting stuck on the variable type lists of strings. I'm guessing that is because the mutation_variables block is a map (of strings) and a list cannot be passed as a single string value to the key pair because it is many values. As an example passing a list of AWS account numbers to my graphql mutation.

Any workaround you can think of for this case?

**Comment by sullivtr** [comment]
Hi @robwolff3. Thanks for asking about this. Due to the way terraform variables work, it is not possible to pass complex objects as the mutation variables. [THIS RESPONSE](https://github.com/sullivtr/terraform-provider-graphql/issues/28#issuecomment-824161468) outlines how one can go about passing values for more complex data structures. 

Now, this becomes a little tricky if the API you are using is forcing you to pass a list as an argument. Do you mind posting what you have so I can help you find a solution?

**Comment by ajbouh** [comment]
I'm also facing this. One option might be to support passing in a JSON string and then parse the JSON before using the values. Here's an example query I'm trying to use for my create mutation over on `https://api.fly.io/graphql`:

```graphql
mutation create(
  $organizationId: ID!
  $name: String
  $config: JSON
  $vmSize: VMSizeEnum
  $regions: [RegionEnum!]
  $count: Int
  $volumes: [VolumeInput!]
  $secrets: [SecretInput!]
  $scheduling: SchedulingEnum
  $image: String!
  $clientMutationId: String
) {
  launchApp(input: {
    organizationId: $organizationId
    name: $name
    config: $config
    vmSize: $vmSize
    regions: $regions
    count: $count
    volumes: $volumes
    secrets: $secrets
    scheduling: $scheduling
    image: $image
    clientMutationId: $clientMutationId
  }) {
    app {
      id
      name
    }
  }
}
```

As you can see I need to specify `$count` as an integer.

I also need to pass in lists for `$regions`, `$secrets`, and `$volumes`.

Trying to inline everything into the graphql query itself is dangerous because it forces a create/destroy which is not necessarily what you want when using this against a production environment.

**Comment by robwolff3** [comment]
Here is my situation trying to pass multiple account numbers into my query. How would you inline @ajbouh ?

main.tf
```terraform
resource "graphql_mutation" "resource" {
  mutation_variables = {
    "title" = var.title
    "description" = var.description
    "defaultImportance" = var.importance
    "accounts" = var.accounts
    "resourceIds" = var.resourceids
    "tags" = var.tags
  }
  read_query_variables = {
    "title" = var.title
  }
  delete_mutation_variables = {
    "srn" = "Resource.srn"
  }
  compute_from_create = true
  compute_mutation_keys = {
    "srn" = "CreateResource.srn"
    "title" = "CreateResource.title"
  }
...
}
```

variables.tf 
```terraform
variable "title" {
  type        = string
}
variable "description" {
  type        = string
}
variable "importance" {
  type        = string
}
variable "accounts" {
  type        = list(string)
  default     = [ "1234567890", "1234567891" ]
}
```

Create Mutation
```graphql
mutation createResource(
  $title: String
  $description: String
  $defaultImportance: Int
  $accounts: [String]
  $resourceIds: [String]
  $tags: [String]
) {
  createResource(value: {
    title: $title
    description: $description
    defaultImportance: $defaultImportance
    accounts: $accounts
    resourceIds: $resourceIds
    tags: $tags
  }){
    srn
    title
  }
}
```

**Comment by ajbouh** [comment]
@robwolff3 use terraform string interpolation and make the number value part of the mutation query itself.

**Comment by sullivtr** [comment]
So after giving this some thought, I believe I can make it possible to do something like this: 

```json
mutation_variables = {
    "someArray" = "[\"string1\", \"string2\"]"
}
```
And then have the query executor unmarshal the json into an object before passing it down to the query. 
You would set the value to a string, but have to hcl escape the string quotes inside the array/object. 

If I can get something working, than the Terraform jsonencode function will be handy for passing more complex data structures to the variables as jsonencode(var.someListOfString)

Y'all will have to give me a few days to work something out since I do not have an API I can test this against, so I need to enhance my e2e test API to handle this. 

Ill keep you up to date :)

**Comment by sullivtr** [comment]
I've made some progress on this. My current solution should work out. I expect to have a PR ready before the end of this weekend. I just need to do extensive testing to ensure backward compat, and no breaking changes. Stay tuned, and thanks again for the feedback.

**Comment by sullivtr** [comment]
@ajbouh @robwolff3 A PR is up for this. I still need to do some manual testing locally to verify a few things, but ultimately this will result in a major version upgrade due to some behavioral changes being breaking

**Comment by sullivtr** [comment]
@ajbouh @robwolff3 I just cut a pre-release version for these fixes on `v2.0.0-rc`. Please give these changes a go, and let me know how things work out, and be sure to report any issues you run into. Thanks! 
Please note that there may be some backward incompatibility depending on how you structured your mutation variables before. You can find updated documentation for these changes here: https://github.com/sullivtr/terraform-provider-graphql/blob/feature/enchance-mutation-variable-inputs-with-json-capability/docs/resources/mutation.md

**Comment by robwolff-sonrai** [comment]
In my initial testing this new feature is working great!
